Separate and refactor SubtitleSync and OffsetController

armanckeser opened the pull request:
<!--
Ensure your title is short, descriptive, and in the imperative mood (Fix X, Change Y, instead of Fixed X, Changed Y).
For a good inspiration of what to write in commit messages and PRs please review https://chris.beams.io/posts/git-commit/ and our https://jellyfin.org/docs/general/contributing/issues page.
-->
Cleans up the implementation, separating offset controller and the subtitle sync feature itself.

**Changes**
<!-- Describe your changes here in 1-5 sentences. -->

**Issues**
<!-- Tag any issues that this PR solves here.
ex. Fixes # -->

This is to pave way for this feature https://github.com/jellyfin/jellyfin-web/pull/6735

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->

## Summary by CodeRabbit

- **Refactor**
  - Improved the subtitle synchronization interface for better modularity and maintainability. No changes to user-facing features or interactions.

<!-- end of auto-generated comment: release notes by coderabbit.ai -->

dmitrylyzo commented on src/components/subtitlesync/subtitlesync.js:
```diff
@@ -1,170 +1,225 @@
+        this.offsetController.adjustOffset(-this.subtitleSyncSlider.step);
+    }
+
+    destroy() {
```
Maybe move `destroy` and `toggle` right after `constructor` to keep the old sequence?
Or do you have any thoughts on their current positions?

armanckeser replied:
Nope, the old sequence sounds fine

dmitrylyzo commented on src/components/subtitlesync/subtitlesync.js:
```diff
@@ -1,170 +1,225 @@
+    }
 
-    instance.element = parent;
+    setOffset(offset) {
+        this.currentOffset = offset;
+        this.slider.value = offset.toString();
+        this.textField.updateOffset(offset);
+    }
```
Alternatively (to `currentOffset` and `setOffset`), we can declare a property with a custom getter and setter.
```js
get currentOffset() {
    return parseFloat(this.slider.value);
}

set currentOffset(offset) {
    this.slider.value = offset.toString();
    this.textField.updateOffset(offset);
}
```

armanckeser replied:
Good call, thanks! I also added the playback manager offset call into it

armanckeser replied:
```
        slider.addEventListener('change', () => {
            // eslint-disable-next-line no-self-assign
            this.currentOffset = this.currentOffset;
        });
```

not a JS native, but I wanted the slider change to trigger setting of the offset so I am assigning the value to itself, is there a better way to do this, or am I missing something

dmitrylyzo replied:
> not a JS native, but I wanted the slider change to trigger setting of the offset so I am assigning the value to itself, is there a better way to do this, or am I missing something (would assigning to `parseFloat(this.slider.value)` be more desirable?)

I'm fine with self-assigning approach. Alternatively, you can extract the playbackManager and textField updates in a separate function, as it was before: https://github.com/jellyfin/jellyfin-web/blob/fec58cc8aaa9dc665105d2677b5b721415360aee/src/components/subtitlesync/subtitlesync.js#L75-L81.

dmitrylyzo reviewed (commented):
Haven't tested, but looks good.

armanckeser commented:
@dmitrylyzo I don't want to nag but wanted to confirm, for reviews do you prefer being tagged or do you just go through your notifications when you have the time?

dmitrylyzo commented on src/components/subtitlesync/subtitlesync.js:
```diff
@@ -1,125 +1,138 @@
+                    inputOffset = parseFloat(inputOffset[0]);
+                    this.currentOffset = inputOffset;
+                } else {
+                    textField.textContent = (playbackManager.getPlayerSubtitleOffset(this.player) || 0) + 's';
```
```suggestion
                    textField.updateOffset(playbackManager.getPlayerSubtitleOffset(this.player) || 0);
```

In the next PR, we need to introduce a new translatable string for offset and use it in `textField.updateOffset`.

dmitrylyzo commented on src/components/subtitlesync/subtitlesync.js:
```diff
@@ -1,125 +1,138 @@
-        updateSubtitleOffset();
-    };
+    set currentOffset(value) {
+        this.slider.value = value.toString();
+        playbackManager.setSubtitleOffset(value, this.player);
```
```suggestion
        this.slider.value = value.toString();

        // rely on slider value trimming
        value = this.currentOffset;

        playbackManager.setSubtitleOffset(value, this.player);
```
Re-assign value to fix floating point rounding errors.
![float-1](https://github.com/user-attachments/assets/048ce16b-7366-4789-9fee-760582bb70cd)
Previously, it was `const value = parseFloat(subtitleSyncSlider.value);`.
You may want to rephrase the comment.

dmitrylyzo reviewed (approved):
LGTM. I think we should use "squash merge".

thornbill commented on src/components/subtitlesync/subtitlesync.js:
```diff
@@ -1,125 +1,142 @@
-    subtitleSyncSlider.updateOffset = function (sliderValue) {
-        // default value is 0s = 0ms
-        this.value = sliderValue === undefined ? 0 : sliderValue;
+    _initSlider() {
```
Small code style feedback: we should use actual private methods instead of the pseudo standard underscore prefix.

Also for future reference, [we don't use "conventional" commits](https://jellyfin.org/docs/general/contributing/development#pull-request-guidelines).

armanckeser replied:
Ah, not a JS native (coming from Python), didn't even know JS had special syntax for private methods, good to learn. I made the change


For the commit, I am used to a squash merge and changed the name of the PR to non conventional, hopefully that is good.

thornbill commented:
Not a blocker but I would suggest removing the underscore from the method names

armanckeser commented:
@dmitrylyzo @thornbill I think this is ready now?

armanckeser commented:
@dmitrylyzo @thornbill, opened another small PR exposing the current track events. If we can get these in I can start working on the subtitle sync view if that's okay. Thank you!

armanckeser commented:
@thornbill hey there! Don't mean to be a thorn on you guys' side but the squeaky wheel gets the grease and I would love to contribute the subtitle sync rendering to this project. I can only imagine how hard it must be to maintain and develop this, so I am happy to accommodate for any preferences, anything that will decrease the merge time for this and the following PRS to days instead of months. Do let me know the conditions that would make you feel comfortable in merging these changes and ultimately the final version #6735. The new PR I opened doesn't use conventional commits and I have squashed it down to a singular commit and plan to do the same for the final commit adding the timeline render.

thornbill commented:
Hey @armanckeser I hate to leave you hanging on this, but I'm just really busy at the moment. On top of my normal responsibilities I am in the process of moving, so really unless something is critical for 10.11 or was work I already had in progress it's not getting the attention it deserves.

Thanks for your patience!

armanckeser commented:
Thanks for the heads up, good luck with the move

Commit:
Separate and refactor SubtitleSync and OffsetController

Extract offset state and interaction handling from SubtitleSync into a
dedicated OffsetController class. Use real private methods, a currentOffset
getter/setter, and a shared slider bubble helper. No user-facing behavior
changes. Paves the way for the subtitle sync timeline feature.

Add Armanc to CONTRIBUTORS.

armanckeser committed:
Merge branch 'master' into armanc/subtitle-sync-refactor

dmitrylyzo reviewed (approved):
Just in case. The changes still look fine. Tested with the current `master` and #8153.

armanckeser committed:
Merge branch 'master' into armanc/subtitle-sync-refactor
